diff --git a/Include/internal/pycore_time.h b/Include/internal/pycore_time.h
index 2331247..e42da26 100644
--- a/Include/internal/pycore_time.h
+++ b/Include/internal/pycore_time.h
@@ -59,7 +59,7 @@ extern "C" {
 
 #include "pycore_runtime_structs.h" // _PyTimeFraction
 
-#ifdef __clang__
+#if defined(__clang__) || defined(__QNX__)
 struct timeval;
 #endif
 
diff --git a/Lib/asyncio/unix_events.py b/Lib/asyncio/unix_events.py
index 1c14581..4468ad5 100644
--- a/Lib/asyncio/unix_events.py
+++ b/Lib/asyncio/unix_events.py
@@ -118,7 +118,8 @@ def add_signal_handler(self, sig, callback, *args):
             signal.signal(sig, _sighandler_noop)
 
             # Set SA_RESTART to limit EINTR occurrences.
-            signal.siginterrupt(sig, False)
+            # Doesn't work on QNX tho
+            #signal.siginterrupt(sig, False)
         except OSError as exc:
             del self._signal_handlers[sig]
             if not self._signal_handlers:
diff --git a/Lib/getpass.py b/Lib/getpass.py
index 3d9bb1f..caae57a 100644
--- a/Lib/getpass.py
+++ b/Lib/getpass.py
@@ -86,6 +86,8 @@ def unix_getpass(prompt='Password: ', stream=None, *, echo_char=None):
                 finally:
                     termios.tcsetattr(fd, tcsetattr_flags, old)
                     stream.flush()  # issue7208
+                    if os.name == 'qnx':
+                        stream.seek(0, 1)
             except termios.error:
                 if passwd is not None:
                     # _raw_input succeeded.  The final tcsetattr failed.  Reraise
diff --git a/Lib/shutil.py b/Lib/shutil.py
index 8d8fe14..4278c40 100644
--- a/Lib/shutil.py
+++ b/Lib/shutil.py
@@ -762,7 +762,7 @@ def _rmtree_safe_fd_step(stack, onexc):
             orig_st = orig_entry.stat(follow_symlinks=False)
 
         func = os.open  # For error reporting.
-        topfd = os.open(name, os.O_RDONLY | os.O_NONBLOCK, dir_fd=dirfd)
+        topfd = os.open(name, os.O_RDONLY | os.O_NONBLOCK | os.O_DIRECTORY, dir_fd=dirfd)
 
         func = os.path.islink  # For error reporting.
         try:
diff --git a/Lib/test/support/os_helper.py b/Lib/test/support/os_helper.py
index 2c45fe2..882253b 100644
--- a/Lib/test/support/os_helper.py
+++ b/Lib/test/support/os_helper.py
@@ -50,7 +50,7 @@
                   % (TESTFN_UNENCODABLE, sys.getfilesystemencoding()))
             TESTFN_UNENCODABLE = None
 # Apple and Emscripten deny unencodable filenames (invalid utf-8)
-elif not support.is_apple and sys.platform not in {"emscripten", "wasi"}:
+elif not support.is_apple and sys.platform not in {"emscripten", "wasi", 'qnx'}:
     try:
         # ascii and utf-8 cannot encode the byte 0xff
         b'\xff'.decode(sys.getfilesystemencoding())
diff --git a/Modules/_localemodule.c b/Modules/_localemodule.c
index 41e6d48..5ff2577 100644
--- a/Modules/_localemodule.c
+++ b/Modules/_localemodule.c
@@ -563,31 +563,31 @@ static struct langinfo_constant{
        OpenBSD doesn't have CODESET but has T_FMT_AMPM, and doesn't have
        a few of the others.
        Solution: ifdef-test them all. */
-#ifdef CODESET
+#if defined(CODESET) || defined(__QNX__)
     LANGINFO(CODESET, LC_CTYPE),
 #endif
-#ifdef T_FMT_AMPM
+#if defined(T_FMT_AMPM) || defined(__QNX__)
     LANGINFO(T_FMT_AMPM, LC_TIME),
 #endif
-#ifdef ERA
+#if defined(ERA) || defined(__QNX__)
     LANGINFO(ERA, LC_TIME),
 #endif
-#ifdef ERA_D_FMT
+#if defined(ERA_D_FMT) || defined(__QNX__)
     LANGINFO(ERA_D_FMT, LC_TIME),
 #endif
-#ifdef ERA_D_T_FMT
+#if defined(ERA_D_T_FMT) || defined(__QNX__)
     LANGINFO(ERA_D_T_FMT, LC_TIME),
 #endif
-#ifdef ERA_T_FMT
+#if defined(ERA_T_FMT) || defined(__QNX__)
     LANGINFO(ERA_T_FMT, LC_TIME),
 #endif
-#ifdef ALT_DIGITS
+#if defined(ALT_DIGITS) || defined(__QNX__)
     LANGINFO(ALT_DIGITS, LC_TIME),
 #endif
-#ifdef YESEXPR
+#if defined(YESEXPR) || defined(__QNX__)
     LANGINFO(YESEXPR, LC_MESSAGES),
 #endif
-#ifdef NOEXPR
+#if defined(NOEXPR) || defined(__QNX__)
     LANGINFO(NOEXPR, LC_MESSAGES),
 #endif
 #ifdef _DATE_FMT
diff --git a/Modules/_testinternalcapi/pytime.c b/Modules/_testinternalcapi/pytime.c
index 2b0a205..41c9aea 100644
--- a/Modules/_testinternalcapi/pytime.c
+++ b/Modules/_testinternalcapi/pytime.c
@@ -2,6 +2,9 @@
 
 #include "parts.h"
 
+#ifdef __QNX__
+#  include <sys/select.h>
+#endif
 #include "pycore_time.h"          // _PyTime_FromSeconds()
 
 #ifdef MS_WINDOWS
diff --git a/Modules/faulthandler.c b/Modules/faulthandler.c
index d49ce79..4c46e2e 100644
--- a/Modules/faulthandler.c
+++ b/Modules/faulthandler.c
@@ -852,7 +852,9 @@ faulthandler_register(int signum, int chain, _Py_sighandler_t *previous_p)
     /* if the signal is received while the kernel is executing a system
        call, try to restart the system call instead of interrupting it and
        return EINTR. */
+#ifndef __QNX__
     action.sa_flags = SA_RESTART;
+#endif
     if (chain) {
         /* do not prevent the signal from being received from within its
            own signal handler */
diff --git a/Modules/posixmodule.c b/Modules/posixmodule.c
index e1576f8..29264bb 100644
--- a/Modules/posixmodule.c
+++ b/Modules/posixmodule.c
@@ -606,6 +606,12 @@ extern char *ctermid_r(char *);
 #  define MODNAME_OBJ &_Py_ID(posix)
 #endif
 
+#ifdef __QNX__
+#  include <unix.h>
+#  include <termios.h>
+#  include <sys/ioctl.h>
+#endif
+
 /*[clinic input]
 # one of the few times we lie about this name!
 module os
@@ -11156,6 +11162,11 @@ os_open_impl(PyObject *module, path_t *path, int flags, int mode, int dir_fd)
 #ifdef HAVE_OPENAT
         if (dir_fd != DEFAULT_DIR_FD) {
             if (HAVE_OPENAT_RUNTIME) {
+#ifdef __QNX__
+		struct stat path_stat = {};
+		fstatat(dir_fd, path->narrow, &path_stat, 0);
+		if(S_ISDIR(path_stat.st_mode)) flags |= O_DIRECTORY;
+#endif
                 fd = openat(dir_fd, path->narrow, flags, mode);
 
             } else {
diff --git a/Modules/selectmodule.c b/Modules/selectmodule.c
index d234d50..ee062e8 100644
--- a/Modules/selectmodule.c
+++ b/Modules/selectmodule.c
@@ -37,6 +37,10 @@
 #undef HAVE_BROKEN_POLL
 #endif
 
+#ifdef __QNX__
+#  include <sys/select.h>
+#endif
+
 /* Windows #defines FD_SETSIZE to 64 if FD_SETSIZE isn't already defined.
    64 is too small (too many people have bumped into that limit).
    Here we boost it.
diff --git a/Modules/termios.c b/Modules/termios.c
index efb5fcc..6597b90 100644
--- a/Modules/termios.c
+++ b/Modules/termios.c
@@ -11,7 +11,9 @@
 
 #include <string.h>               // memset()
 #include <sys/ioctl.h>
-#include <termios.h>
+#ifndef __QNX__
+#  include <termios.h>
+#endif
 #include <unistd.h>               // _POSIX_VDISABLE
 
 // On QNX 6, struct termio must be declared by including sys/termio.h
diff --git a/Objects/mimalloc/prim/unix/prim.c b/Objects/mimalloc/prim/unix/prim.c
index 1598eba..24b00f9 100644
--- a/Objects/mimalloc/prim/unix/prim.c
+++ b/Objects/mimalloc/prim/unix/prim.c
@@ -51,7 +51,7 @@ terms of the MIT license. A copy of the license can be found in the file
   #include <sys/sysctl.h>
 #endif
 
-#if !defined(__HAIKU__) && !defined(__APPLE__) && !defined(__CYGWIN__) && !defined(_AIX) && !defined(__OpenBSD__) && !defined(__FreeBSD__) && !defined(__sun) && !defined(__NetBSD__)
+#if !defined(__HAIKU__) && !defined(__APPLE__) && !defined(__CYGWIN__) && !defined(_AIX) && !defined(__OpenBSD__) && !defined(__FreeBSD__) && !defined(__sun) && !defined(__NetBSD__) && !defined(__QNX__)
   #define MI_HAS_SYSCALL_H
   #include <sys/syscall.h>
 #endif
@@ -156,6 +156,8 @@ int _mi_prim_free(void* addr, size_t size ) {
 static int unix_madvise(void* addr, size_t size, int advice) {
   #if defined(__sun)
   return madvise((caddr_t)addr, size, advice);  // Solaris needs cast (issue #520)
+  #elif defined(__QNX__)
+  return posix_madvise(addr, size, advice);
   #else
   return madvise(addr, size, advice);
   #endif
@@ -370,7 +372,11 @@ int _mi_prim_commit(void* start, size_t size, bool* is_zero) {
 int _mi_prim_decommit(void* start, size_t size, bool* needs_recommit) {
   int err = 0;
   // decommit: use MADV_DONTNEED as it decreases rss immediately (unlike MADV_FREE)
+  #ifdef __QNX__
+  err = unix_madvise(start, size, POSIX_MADV_DONTNEED);
+  #else
   err = unix_madvise(start, size, MADV_DONTNEED);
+  #endif
   #if !MI_DEBUG && !MI_SECURE
     *needs_recommit = false;
   #else
@@ -402,6 +408,8 @@ int _mi_prim_reset(void* start, size_t size) {
     mi_atomic_store_release(&advice, (size_t)MADV_DONTNEED);
     err = unix_madvise(start, size, MADV_DONTNEED);
   }
+  #elif defined(__QNX__)
+  int err = unix_madvise(start, size, POSIX_MADV_DONTNEED);
   #else
   int err = unix_madvise(start, size, MADV_DONTNEED);
   #endif
diff --git a/Python/fileutils.c b/Python/fileutils.c
index 78603d4..6f5522f 100644
--- a/Python/fileutils.c
+++ b/Python/fileutils.c
@@ -1643,6 +1643,11 @@ _Py_open_impl(const char *pathname, int flags, int gil_held)
 
         do {
             Py_BEGIN_ALLOW_THREADS
+            #ifdef __QNX__
+            struct stat path_stat = {};
+            stat(pathname, &path_stat);
+            if(S_ISDIR(path_stat.st_mode)) flags |= O_DIRECTORY;
+            #endif
             fd = open(pathname, flags);
             Py_END_ALLOW_THREADS
         } while (fd < 0
@@ -1659,6 +1664,11 @@ _Py_open_impl(const char *pathname, int flags, int gil_held)
         Py_DECREF(pathname_obj);
     }
     else {
+        #ifdef __QNX__
+        struct stat path_stat = {};
+        stat(pathname, &path_stat);
+        if(S_ISDIR(path_stat.st_mode)) flags |= O_DIRECTORY;
+        #endif
         fd = open(pathname, flags);
         if (fd < 0)
             return -1;
diff --git a/configure b/configure
index d31c24d..dc82353 100755
--- a/configure
+++ b/configure
@@ -7546,7 +7546,7 @@ then
 	QNX*)
 	   # qcc must be used because the other compilers do not
 	   # support -N.
-	   LINKCC=qcc;;
+	   LINKCC=gcc;;
 	esac
 fi
 { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $LINKCC" >&5
@@ -7732,6 +7732,16 @@ printf "%s\n" "#define Py_ENABLE_SHARED 1" >>confdefs.h
           INSTSONAME="$LDLIBRARY".$SOVERSION
       fi
 
+      if test "$with_pydebug" != yes
+      then
+        PY3LIBRARY=libpython3.so
+      fi
+      ;;
+    QNX)
+      LDLIBRARY='libpython$(LDVERSION).so'
+      BLDLIBRARY='-L. -lpython$(LDVERSION)'
+      RUNSHARED=LD_LIBRARY_PATH=`pwd`${LD_LIBRARY_PATH:+:${LD_LIBRARY_PATH}}
+
       if test "$with_pydebug" != yes
       then
         PY3LIBRARY=libpython3.so
@@ -13757,7 +13767,7 @@ printf "%s\n" "#define THREAD_STACK_SIZE 0x$stack_size" >>confdefs.h
 		# is loaded.  -N 2048K causes the stack size to be set
 		# to 2048 kilobytes so that the stack doesn't overflow
 		# when running test_compile.py.
-		LINKFORSHARED='-Wl,-E -N 2048K';;
+		LINKFORSHARED='-Wl,-export-dynamic -Wl,-E -Wl,--stack,8388608';;
 	VxWorks*)
 		LINKFORSHARED='-Wl,-export-dynamic';;
 	esac
